File: test/hello_world.jl
using LowRankModels, DataFrames, Random, SparseArrays
Random.seed!(0)

# loss types to test
real_loss_types = [QuadLoss, HuberLoss]
bool_loss_types = [HingeLoss]
ordinal_loss_types = [OrdinalHingeLoss, BvSLoss]
categorical_loss_types = [MultinomialLoss, OvALoss]

#instantiate losses
ncat = 4 # maximum categorical levels
nord = 5 # maximum ordinal levels
real_losses = [l() for l in real_loss_types]
bool_losses = [l() for l in bool_loss_types]
ordinal_losses = [l(rand(3:nord)) for l in ordinal_loss_types]
categorical_losses = [l(rand(3:ncat)) for l in categorical_loss_types]
losses = [real_losses..., bool_losses..., ordinal_losses..., categorical_losses...]
data_types = cat([:real for l in real_losses],
                  [:bool for l in bool_losses],
                  [:ord for l in ordinal_losses],
                  [:cat for l in categorical_losses],
                  dims=1)

# scale losses for different columns
for loss in losses
  mul!(loss, rand())
end

# regularizers to test
regularizers = [QuadReg(), OneReg(5), NonNegConstraint(), KSparseConstraint(2)]
# add more regularizers = more rows so the data isn't degenerate
regularizers = cat(regularizers, fill(QuadReg(), 10), dims=1)

m,n = length(regularizers), length(losses)

A_real = rand(m, length(real_losses))
A_bool = rand(Bool, m, length(bool_losses))
A_ord = rand(1:5, m, length(ordinal_losses))
A_cat = rand(1:3, m, length(categorical_losses))

# without saying "Any", upconverts to array of Floats
A = Any[A_real A_bool A_ord A_cat]

glrm = GLRM(A, losses, regularizers, QuadReg(), 2)
fit!(glrm, verbose=false)
println("successfully fit matrix")

Ω = [(rand(1:m), rand(1:n)) for iobs in 1:(5*max(m,n))] # observe some random entries, with replacement
glrm = GLRM(A, losses, regularizers, QuadReg(), 2, obs=Ω);
fit!(glrm, verbose=false)
println("successfully fit matrix with some entries unobserved")

### now fit data frame
A_sparse = sprandn(10, 10, .5)
df = NaNs_to_Missing!(DataFrame(Array(0 ./ A_sparse + A_sparse)))
# explicitly encoding missing
obs = observations(df)
glrm = GLRM(df, QuadLoss(), QuadReg(), QuadReg(), 2, obs=obs)
fit!(glrm, verbose=false)

# implicitly encoding missings from dataframe - this functionality has not been implemented for dataframes
# glrm = GLRM(df, QuadLoss(), QuadReg(), QuadReg(), 2)
# fit!(glrm, verbose=false)

# without specifying losses directly
glrm = GLRM(DataFrame(A), 3, data_types)
fit!(glrm, verbose=false)
println("successfully fit dataframe")

### imputation and sampling
impute(glrm)
println("successfully imputed entries")
sample(glrm)
sample_missing(glrm)
println("successfully sampled from model")

### now fit sparse matrix

m, n = 10, 10
sparseA = sprandn(m, n, .5)
glrm = GLRM(A, QuadLoss(), QuadReg(), QuadReg(), 5)
fit!(glrm, verbose=false)
println("successfully fit sparse GLRM")

File: python/hello_world.py
from numpy.random import rand, randn, choice
from numpy import hstack

# from LowRankModels import * # import julia functions into python
import julia
j = julia.Julia()
j.using("LowRankModels")

# from julia import LowRankModels as lrm

# loss types to test - all these are Julia functions imported from LowRankModels
real_loss_types = [j.QuadLoss, j.HuberLoss]
bool_loss_types = [j.HingeLoss]
ordinal_loss_types = [j.OrdinalHingeLoss, j.BvSLoss]
categorical_loss_types = [j.MultinomialLoss, j.OvALoss]

#instantiate losses
ncat = 4 # maximum categorical levels
nord = 5 # maximum ordinal levels
real_losses = [l() for l in real_loss_types] # a python list of julia objects formed by calling julia functions
bool_losses = [l() for l in bool_loss_types]
ordinal_losses = [l(choice(range(3,nord))) for l in ordinal_loss_types]
categorical_losses = [l(choice(range(3,ncat))) for l in categorical_loss_types]
losses = real_losses + bool_losses + ordinal_losses + categorical_losses # concatenate python lists of julia objects

# scale losses for different columns
for loss in losses:
  j.scale_b(loss, rand()) # scale_b is the julia function scale!

# regularizers to test
regularizers = [j.QuadReg(), j.OneReg(5), j.NonNegConstraint(), j.KSparseConstraint(2)] # python list of julia objects

m = len(regularizers)
n = len(losses)

# matrices of data corresponding to each loss type
A_real = rand(m, len(real_losses))
A_bool = choice([True, False], size = (m, len(bool_losses)))
A_ord = choice(range(1,6), size = (m, len(ordinal_losses))) # use 1 indexing
A_cat = choice(range(1,4), size = (m, len(categorical_losses))) # use 1 indexing

# horizontally concatenate to form one big data matrix
A = hstack([A_real, A_bool, A_ord, A_cat])

# now call main function
# tricky business:
    # * A is a numpy array, will need to be cast as Julia array
    # * losses is a python list of julia objects, will need to be cast as a Julia Array{Loss,1}
    # * regularizers is a python list of julia objects, will need to be cast as a Julia Array{Regularizer,1}
    # * QuadReg() forms a julia object, no problem
    # * 2 is a python int, will need to be cast as a julia int
glrm = j.GLRM(A, losses, regularizers, j.QuadReg(), 2) # GLRM is a julia function from LowRankModels
j.fit_b(glrm) # fit_b is the julia function fit!
print("successfully fit matrix")

### now fit data frame
j.using("DataFrames")
# could also try a pandas table instead...?
A = j.DataFrame(A)
for i in range(10):
  A[choice(range(1,m+1)), choice(range(1,n+1))] = j.NA
obs = j.observations(A)
glrm = j.GLRM(A, losses, QuadReg(), QuadReg(), 2, obs=obs)
j.fit_b(glrm)
print("successfully fit dataframe")

# imputation and sampling
j.impute(glrm)
print("successfully imputed entries")
j.sample(glrm)
j.sample_missing(glrm)
print("successfully sampled from model")

### now fit sparse matrix
from scipy.sparse import random

m, n = 100, 200
A = random(m, n, .5, format='csc')
glrm = j.GLRM(A, QuadLoss(), QuadReg(), QuadReg(), 5)
j.fit_b(glrm)
print("successfully fit csc sparse GLRM")

A = random(m, n, .5, format='coo')
glrm = j.GLRM(A, QuadLoss(), QuadReg(), QuadReg(), 5)
j.fit_b(glrm)
print("successfully fit coo sparse GLRM")

A = random(m, n, .5, format='csr')
glrm = j.GLRM(A, QuadLoss(), QuadReg(), QuadReg(), 5)
j.fit_b(glrm)
print("successfully fit csr sparse GLRM")

### the hardest test: what if someone passes a pandas dataframe?
import pandas as pd
import numpy as np
df2 = pd.DataFrame({ 'A' : 1.,
                      'B' : pd.Timestamp('20130102'),
                      'C' : pd.Series(1,index=list(range(4)),dtype='float32'),
                      'D' : np.array([3] * 4,dtype='int32'),
                      'E' : pd.Categorical(["test","train","test","train"]),
                      'F' : 'foo' })

# todo: finish writing this test and fit a GLRM on the pandas dataframe!
